app/models/sendgrid_event.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+class SendgridEvent < ApplicationRecord
+
+  store_accessor :payload, :email
+
+  # To make allowances for occasional inbox down time, this counts a maximum of one fail per day,
+  # e.g.:
+  #
+  # - If 2 emails are sent on the same day, and both fail, then this counts as only 1 fail, not
+  #   2 fails.
+  #
+  # - If 2 emails are sent on 2 separate days, and both fail, then this counts as 2
+  #   fails, as the failures were on different days.
+  def self.fails_since_last_delivery(email)
+    last_delivered_at = where_email(email).where_event_type("delivered").maximum(timestamp_sql)
+
+    fails_query =
+      select("DISTINCT(date_trunc('day', #{timestamp_sql}))").where_email(email).where_event_type("dropped", "bounce")
+
+    fails_query = fails_query.where("#{timestamp_sql} > ?", last_delivered_at) if last_delivered_at
+
+    fails_query.count
+  end
+
+  def self.where_email(email)
+    where("(payload->>'email') = ?", email)
+  end
+
+  def self.where_event_type(*event_types)
+    where("(payload->>'event') IN (?)", event_types)
+  end
+
+  def self.where_sg_event_id(sg_event_id)
```

Comment:
I think we don't need to use `sg` instead of `sendgird` to keep code easy understandable.

Reply:
Latest commit includes dedicated column `sendgrid_events.sendgrid_id` to store `sg_event_id`.